widget/completionentry.go

```diff
@@ -63,168 +69,202 @@ func (c *CompletionEntry) ShowCompletion() {
+		parent.selected = id
+	}
+	list.List.OnUnselected = func(_ widget.ListItemID) {
+		parent.selected = -1
+	}
+	return list
 }
 
 // Implements: fyne.Focusable
-func (n *navigableList) FocusLost() {
-}
+func (list *completionEntryList) FocusGained() {}
 
-func (n *navigableList) SetOptions(items []string) {
-	n.Unselect(n.selected)
-	n.items = items
-	n.Refresh()
-	n.selected = -1
-}
+// Implements: fyne.Focusable
+func (list *completionEntryList) FocusLost() {}
 
-func (n *navigableList) TypedKey(event *fyne.KeyEvent) {
-	switch event.Name {
+// Implements: fyne.Focusable
+func (list *completionEntryList) TypedKey(ke *fyne.KeyEvent) {
+	switch ke.Name {
 	case fyne.KeyDown:
-		if n.selected < len(n.items)-1 {
-			n.selected++
+		if list.parent.selected < len(list.parent.Options)-1 {
+			list.parent.list.Select(list.parent.selected + 1)
 		} else {
-			n.selected = 0
+			list.parent.list.Select(0)
 		}
-		n.navigating = true
-		n.Select(n.selected)
-
 	case fyne.KeyUp:
-		if n.selected > 0 {
-			n.selected--
+		if list.parent.selected > 0 {
+			list.parent.list.Select(list.parent.selected - 1)
 		} else {
-			n.selected = len(n.items) - 1
+			list.parent.list.Select(len(list.parent.Options) - 1)
 		}
-		n.navigating = true
-		n.Select(n.selected)
 	case fyne.KeyReturn, fyne.KeyEnter:
-		if n.selected == -1 { // so the user want to submit the entry
-			n.hide()
-			n.entry.TypedKey(event)
+		if list.parent.selected >= 0 {
+			list.parent.setTextFromMenu(list.parent.Options[list.parent.selected])
 		} else {
-			n.navigating = false
-			n.OnSelected(n.selected)
+			list.parent.HideCompletion()
+			list.parent.Entry.TypedKey(ke)
 		}
-	case fyne.KeyEscape:
-		n.hide()
+	case fyne.KeyTab, fyne.KeyEscape:
+		list.parent.HideCompletion()
 	default:
-		n.entry.TypedKey(event)
-
+		list.parent.TypedKey(ke)
 	}
 }
 
-func (n *navigableList) TypedRune(r rune) {
-	n.entry.TypedRune(r)
+// Implements: fyne.Focusable
+func (list *completionEntryList) TypedRune(r rune) {
+	list.parent.TypedRune(r)
+}
+
+// Implements: fyne.Shortcutable
+func (list *completionEntryList) TypedShortcut(s fyne.Shortcut) {
+	list.parent.TypedShortcut(s)
+}
+
+type completionEntryListItem struct {
+	widget.BaseWidget
+	parent *CompletionEntry
+	id     widget.ListItemID // each list item knows where it belongs in parent's data slice
+	co     fyne.CanvasObject
+}
+
+func newCompletionEntryListItem(parent *CompletionEntry, co fyne.CanvasObject) *completionEntryListItem {
+	item := &completionEntryListItem{parent: parent, id: -1, co: co}
+	item.ExtendBaseWidget(item)
+	return item
 }
+
+func (item *completionEntryListItem) CreateRenderer() fyne.WidgetRenderer {
+	return widget.NewSimpleRenderer(item.co)
+}
+
+func (item *completionEntryListItem) Tapped(_ *fyne.PointEvent) {
+	item.parent.setTextFromMenu(item.parent.Options[item.id])
+}
+
+func (item *completionEntryListItem) MouseIn(_ *desktop.MouseEvent)    { item.parent.list.Select(item.id) }
```

Comment:
I think it's not OK that this only happens on desktop instance.
I can't see how mobile apps will be able to select an item.

Reply:
This will just select an item to show it selected on hover, item completion is not triggered by list.Select, but by item Tapped